### DISCUSSION TITLE
How to run several parallel workflow instances on different threads?

### DISCUSSION THREAD

**Discussion by Invenisso** [elsa-workflows/elsa-core#2444] [Q&A]
Hello Community!

I slightly modified Elsa.Samples.CustomOutcomeActivityConsole example to see if I would be able to run 10 parallel workflow instances in separate threads:

![image](https://user-images.githubusercontent.com/57659054/139867891-8de9c304-87d0-405a-8a77-828123f474c7.png)

But I am getting the error: 

![image](https://user-images.githubusercontent.com/57659054/139867558-56104659-c3db-42c5-a69a-cbf391b54a5b.png)

So, what is the correct way to run several parallel workflow instances? I mean truly parallel ones - actually performing side by side on different threads.

**Comment by sfmskywalker** [comment]
You should be able to run workflow instances in parallel, but as you can see there is an issue with the `InMemoryStore<T>` implementation. Should be fixed easily by using a semaphore and/or using a concurrent dictionary.

I'll create an issue for this.

**Reply by sfmskywalker** [reply]
Related issue: #2483
